File: controllers/authController.js
/**
 * controllers/authController.js
 * Handles login, registration, current-user info, and user registry.
 */

const jwt = require('jsonwebtoken');
const User = require('../models/userModel');
const { JWT_SECRET } = require('../middleware/auth');

// ─── POST /api/auth/login ─────────────────────────────────────────────────────

const login = (req, res) => {
  const { username, password } = req.body;

  if (!username || !password) {
    return res.status(400).json({ success: false, error: 'Username and password are required.' });
  }

  const user = User.findByUsername(username.trim());
  if (!user || user.password !== password) {
    return res.status(401).json({ success: false, error: 'Invalid credentials.' });
  }

  const payload = { id: user.id, username: user.username, name: user.name, role: user.role };
  const token = jwt.sign(payload, JWT_SECRET, { expiresIn: '24h' });

  res.json({ success: true, token, user: payload });
};

// ─── POST /api/auth/register ──────────────────────────────────────────────────

const register = (req, res) => {
  const { username, password, name } = req.body;

  const errors = [];
  if (!username || username.trim().length < 3) errors.push('"username" must be at least 3 characters.');
  if (!password || password.length < 4) errors.push('"password" must be at least 4 characters.');
  if (!name || name.trim().length < 2) errors.push('"name" must be at least 2 characters.');
  if (errors.length) {
    return res.status(400).json({ success: false, error: 'Validation failed.', details: errors });
  }

  if (User.findByUsername(username.trim())) {
    return res.status(409).json({ success: false, error: `Username "${username.trim()}" is already taken.` });
  }

  const newUser = User.create({ username, password, name });

  const payload = { id: newUser.id, username: newUser.username, name: newUser.name, role: newUser.role };
  const token = jwt.sign(payload, JWT_SECRET, { expiresIn: '24h' });

  res.status(201).json({ success: true, token, user: payload });
};

// ─── GET /api/auth/me ─────────────────────────────────────────────────────────

const me = (req, res) => {
  // Return full user data (not just JWT payload) so avatarUrl etc. are included
  const full = User.findById(req.user.id);
  if (!full) return res.status(404).json({ success: false, error: 'User not found.' });
  const { password: _pw, ...publicUser } = full;
  res.json({ success: true, user: publicUser });
};

// ─── GET /api/auth/users ─────────────────────────────────────────────────────
// Admin only — returns all registered users (no passwords)

const getUsers = (_req, res) => {
  const users = User.getAll();
  res.json({ success: true, count: users.length, data: users });
};

// ─── PUT /api/auth/users/:id ──────────────────────────────────────────────────
// Admin only — full update of a user

const updateUser = (req, res) => {
  const { id } = req.params;
  const { name, username, password, role } = req.body;

  const existing = User.findById(id);
  if (!existing) return res.status(404).json({ success: false, error: `User #${id} not found.` });

  const errors = [];
  if (name !== undefined && name.trim().length < 2) errors.push('"name" must be at least 2 characters.');
  if (username !== undefined && username.trim().length < 3) errors.push('"username" must be at least 3 characters.');
  if (password !== undefined && password.length < 4) errors.push('"password" must be at least 4 characters.');
  if (role !== undefined && !['admin', 'user'].includes(role)) errors.push('"role" must be "admin" or "user".');
  if (errors.length) return res.status(400).json({ success: false, error: 'Validation failed.', details: errors });

  // Check username uniqueness if changing it
  if (username && username.trim() !== existing.username) {
    const taken = User.findByUsername(username.trim());
    if (taken) return res.status(409).json({ success: false, error: `Username "${username.trim()}" is already taken.` });
  }

  const updated = User.update(id, { name, username, password, role });
  res.json({ success: true, data: updated });
};

// ─── PATCH /api/auth/users/:id ────────────────────────────────────────────────
// Admin only — partial update; also allows a user to edit their own profile

const patchUser = (req, res) => {
  const { id } = req.params;
  const fields = req.body;

  const existing = User.findById(id);
  if (!existing) return res.status(404).json({ success: false, error: `User #${id} not found.` });

  // Non-admins can only edit themselves, and cannot change their role
  if (req.user.role !== 'admin') {
    if (req.user.id !== Number(id)) {
      return res.status(403).json({ success: false, error: 'You can only edit your own profile.' });
    }
    delete fields.role; // users cannot elevate their own role
  }

  const errors = [];
  if (fields.name !== undefined && fields.name.trim().length < 2) errors.push('"name" must be at least 2 characters.');
  if (fields.username !== undefined && fields.username.trim().length < 3) errors.push('"username" must be at least 3 characters.');
  if (fields.password !== undefined && fields.password.length < 4) errors.push('"password" must be at least 4 characters.');
  if (fields.role !== undefined && !['admin', 'user'].includes(fields.role)) errors.push('"role" must be "admin" or "user".');
  if (errors.length) return res.status(400).json({ success: false, error: 'Validation failed.', details: errors });

  if (fields.username && fields.username.trim() !== existing.username) {
    const taken = User.findByUsername(fields.username.trim());
    if (taken) return res.status(409).json({ success: false, error: `Username "${fields.username.trim()}" is already taken.` });
  }

  const patched = User.patch(id, fields);
  res.json({ success: true, data: patched });
};

// ─── DELETE /api/auth/users/:id ───────────────────────────────────────────────
// Admin only — remove a user account

const deleteUser = (req, res) => {
  const { id } = req.params;

  if (!User.findById(id)) {
    return res.status(404).json({ success: false, error: `User #${id} not found.` });
  }
  if (req.user.id === Number(id)) {
    return res.status(400).json({ success: false, error: 'You cannot delete your own account.' });
  }

  User.remove(id);
  res.json({ success: true, message: `User #${id} deleted.` });
};

module.exports = { login, register, me, getUsers, updateUser, patchUser, deleteUser };

File: src/main/java/com/sportsupps/store/controller/AuthController.java
package com.sportsupps.store.controller;

import com.sportsupps.store.dto.ApiResponse;
import com.sportsupps.store.dto.LoginRequest;
import com.sportsupps.store.dto.RegisterRequest;
import com.sportsupps.store.model.User;
import com.sportsupps.store.security.AuthUser;
import com.sportsupps.store.security.JwtUtil;
import com.sportsupps.store.service.UserService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final UserService userService;
    private final JwtUtil jwtUtil;

    public AuthController(UserService userService, JwtUtil jwtUtil) {
        this.userService = userService;
        this.jwtUtil = jwtUtil;
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody LoginRequest req) {
        if (req.getUsername() == null || req.getPassword() == null)
            return ResponseEntity.badRequest().body(ApiResponse.error("Username and password are required."));
        return userService.findByUsername(req.getUsername().trim())
                .filter(u -> u.getPassword().equals(req.getPassword()))
                .map(u -> {
                    String token = jwtUtil.generateToken(u.getId(), u.getUsername(), u.getName(), u.getRole());
                    Map<String, Object> payload = Map.of("id", u.getId(), "username", u.getUsername(), "name", u.getName(), "role", u.getRole());
                    return ResponseEntity.ok(ApiResponse.builder().success(true).token(token).user(payload).build());
                })
                .orElseGet(() -> ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(ApiResponse.error("Invalid credentials.")));
    }

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody RegisterRequest req) {
        if (req.getUsername() == null || req.getUsername().trim().length() < 3)
            return ResponseEntity.badRequest().body(ApiResponse.error("\"username\" must be at least 3 characters."));
        if (req.getPassword() == null || req.getPassword().length() < 4)
            return ResponseEntity.badRequest().body(ApiResponse.error("\"password\" must be at least 4 characters."));
        if (req.getName() == null || req.getName().trim().length() < 2)
            return ResponseEntity.badRequest().body(ApiResponse.error("\"name\" must be at least 2 characters."));
        if (userService.existsByUsername(req.getUsername().trim()))
            return ResponseEntity.status(HttpStatus.CONFLICT).body(ApiResponse.error("Username \"" + req.getUsername().trim() + "\" is already taken."));

        User newUser = User.builder()
                .username(req.getUsername().trim()).password(req.getPassword())
                .name(req.getName().trim()).role("user")
                .avatarUrl(req.getAvatarUrl() != null ? req.getAvatarUrl() : "")
                .build();
        newUser = userService.save(newUser);
        String token = jwtUtil.generateToken(newUser.getId(), newUser.getUsername(), newUser.getName(), newUser.getRole());
        Map<String, Object> payload = Map.of("id", newUser.getId(), "username", newUser.getUsername(), "name", newUser.getName(), "role", newUser.getRole());
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.builder().success(true).token(token).user(payload).build());
    }

    @GetMapping("/me")
    public ResponseEntity<?> me(@AuthenticationPrincipal AuthUser authUser) {
        return userService.findById(authUser.getId())
                .map(u -> ResponseEntity.ok(ApiResponse.builder().success(true).user(toPublicUser(u)).build()))
                .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND).body(ApiResponse.error("User not found.")));
    }

    @GetMapping("/users")
    public ResponseEntity<?> getUsers() {
        List<Map<String, Object>> users = userService.findAll().stream().map(this::toPublicUser).toList();
        return ResponseEntity.ok(ApiResponse.builder().success(true).count(users.size()).data(users).build());
    }

    @PutMapping("/users/{id}")
    public ResponseEntity<?> updateUser(@PathVariable Integer id, @RequestBody Map<String, Object> body) {
        return userService.findById(id).map(u -> {
            String name     = body.containsKey("name")      ? (String) body.get("name")      : u.getName();
            String username = body.containsKey("username")  ? (String) body.get("username")  : u.getUsername();
            String password = body.containsKey("password")  ? (String) body.get("password")  : u.getPassword();
            String role     = body.containsKey("role")      ? (String) body.get("role")      : u.getRole();
            String avatar   = body.containsKey("avatarUrl") ? (String) body.get("avatarUrl") : u.getAvatarUrl();

            if (name != null && name.trim().length() < 2)     return ResponseEntity.badRequest().body(ApiResponse.error("\"name\" must be at least 2 characters."));
            if (username != null && username.trim().length() < 3) return ResponseEntity.badRequest().body(ApiResponse.error("\"username\" must be at least 3 characters."));
            if (password != null && password.length() < 4)    return ResponseEntity.badRequest().body(ApiResponse.error("\"password\" must be at least 4 characters."));
            if (role != null && !role.equals("admin") && !role.equals("user")) return ResponseEntity.badRequest().body(ApiResponse.error("\"role\" must be \"admin\" or \"user\"."));
            if (!username.equals(u.getUsername()) && userService.existsByUsername(username)) return ResponseEntity.status(HttpStatus.CONFLICT).body(ApiResponse.error("Username \"" + username + "\" is already taken."));

            u.setName(name.trim()); u.setUsername(username.trim()); u.setPassword(password);
            u.setRole(role); u.setAvatarUrl(avatar != null ? avatar : "");
            return ResponseEntity.ok(ApiResponse.ok(toPublicUser(userService.save(u))));
        }).orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND).body(ApiResponse.error("User #" + id + " not found.")));
    }

    @PatchMapping("/users/{id}")
    public ResponseEntity<?> patchUser(@PathVariable Integer id, @RequestBody Map<String, Object> body,
                                       @AuthenticationPrincipal AuthUser authUser) {
        return userService.findById(id).map(u -> {
            if (!authUser.isAdmin() && !authUser.getId().equals(id))
                return ResponseEntity.status(HttpStatus.FORBIDDEN).body(ApiResponse.error("You can only edit your own profile."));
            if (!authUser.isAdmin()) body.remove("role");

            if (body.containsKey("name"))      { String v = ((String) body.get("name")).trim();     if (v.length() < 2) return ResponseEntity.badRequest().body(ApiResponse.error("\"name\" must be at least 2 characters.")); u.setName(v); }
            if (body.containsKey("username"))  { String v = ((String) body.get("username")).trim(); if (v.length() < 3) return ResponseEntity.badRequest().body(ApiResponse.error("\"username\" must be at least 3 characters.")); if (!v.equals(u.getUsername()) && userService.existsByUsername(v)) return ResponseEntity.status(HttpStatus.CONFLICT).body(ApiResponse.error("Username \"" + v + "\" is already taken.")); u.setUsername(v); }
            if (body.containsKey("password"))  { String v = (String) body.get("password"); if (v.length() < 4) return ResponseEntity.badRequest().body(ApiResponse.error("\"password\" must be at least 4 characters.")); u.setPassword(v); }
            if (body.containsKey("role"))      { String v = (String) body.get("role"); if (!v.equals("admin") && !v.equals("user")) return ResponseEntity.badRequest().body(ApiResponse.error("\"role\" must be \"admin\" or \"user\".")); u.setRole(v); }
            if (body.containsKey("avatarUrl")) u.setAvatarUrl((String) body.get("avatarUrl"));

            return ResponseEntity.ok(ApiResponse.ok(toPublicUser(userService.save(u))));
        }).orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND).body(ApiResponse.error("User #" + id + " not found.")));
    }

    @DeleteMapping("/users/{id}")
    public ResponseEntity<?> deleteUser(@PathVariable Integer id, @AuthenticationPrincipal AuthUser authUser) {
        if (!userService.findById(id).isPresent())
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ApiResponse.error("User #" + id + " not found."));
        if (authUser.getId().equals(id))
            return ResponseEntity.badRequest().body(ApiResponse.error("You cannot delete your own account."));
        userService.deleteById(id);
        return ResponseEntity.ok(ApiResponse.deleted("User #" + id + " deleted."));
    }

    private Map<String, Object> toPublicUser(User u) {
        Map<String, Object> m = new HashMap<>();
        m.put("id", u.getId()); m.put("username", u.getUsername()); m.put("name", u.getName());
        m.put("role", u.getRole()); m.put("avatarUrl", u.getAvatarUrl()); m.put("createdAt", u.getCreatedAt());
        return m;
    }
}
